text/0000-nested-publish.md

```diff
@@ -0,0 +1,198 @@
+* If your package contains `[dev-dependencies]` which do not give a `version = "..."`, they are stripped out on publication.
+
+You can change this default by placing in the manifest (`Cargo.toml`) of a sub-package:
+
+```toml
+[package]
+publish = "nested"
+```
+
+If this is done, Cargo's behavior changes as follows:
+
+* If you publish the parent package, the sub-package is included in the `.crate` file (unless overridden by explicit `exclude`/`include`) and will be available to the parent package whenever the parent package is downloaded and compiled.
+* The parent package (and other sub-packages) may have `path =` dependencies upon the sub-package. (Such dependencies must not have a `version =` or `git =`; that is, the `path` must be the _only_ source for the dependency.)
+* You cannot `cargo publish` the sub-package, just as if it had `publish = false`. (This is a safety measure against accidentally publishing the sub-package separately when this is not intended.)
+
+Nested sub-packages may be freely placed within other nested sub-packages.
+
+When a group of packages is published in this way, and depended on, this has a number of useful effects (which are not things that Cargo explicitly implements, just consequences of the system):
+
+* The packages are a single unit for all versioning purposes; there is no way for a version mismatch to arise since all the code was published together. Version resolution does not apply (in the same way that it does not for any other `path =` dependency).
+* The sub-package is effectively “private” (in a sense like the Rust language's visibility system): it cannot be named as a dependency by any other package on `crates.io`, only by its parent package and sibling sub-packages. The parent package may still re-export items from it, or even the entire crate, in the same ways as it could do with a dependency on a normally published package.
+
+## Example: trait and derive macro
+
+Suppose we want to declare a trait-and-derive-macro package. We can do this as follows. The parent package would have this manifest `foo/Cargo.toml`:
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2021"
+publish = true
+
+[dependencies]
+foo-macros = { path = "macros" }    # newly permitted
+```
+
+The sub-package manifest `foo/macros/Cargo.toml`:
+
+```toml
+[package]
+name = "macros"                     # this name need not be claimed on crates.io
+version = "0.1.0"                   # this version is not used for dependency resolution
+edition = "2021"
+publish = "nested"                  # new syntax
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent `foo` directory, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
```

Comment:
I'm missing a section on how we want to make this work with the [crate index](https://github.com/rust-lang/crates.io-index).

We could just skip the all dependencies to a subpackage, but then the dependencies of that subpackage would be missed by the resolver.

If we add some new structure inside the index format to support subpackages then we will need to ensure it doesn't break older cargo implementations.

Reply:
They key part here for cargo's resolver is that it operates on the index with a `Summary` for each package, without downloading the `.crate` files.  This is a way the proposal is different than traditional path dependencies: normal path dependencies are expected to always be there without any kind of `Summary` while anything associated with the registry is not.

Even if this wasn't a problem for cargo, I think that'd be bad for the Index to have holes where it says "trust somthing else to fill in this hole" but that it should be self-consistent.

Reply:
Been thinking about this more.

I think the way to approach this is: when running `cargo package`, flatten nested package features and dependencies into the current package starting at the deepest dependency and recursing up to the selected package.  When `cargo publish` uploads a `Summary`, we will generate it from this flattened set of dependencies, ignoring the nested dependencies that are still left.  This would mean that the package in the Index shows up as if there are no nested packages.  We then resolve dependencies without consideration for the nested packages (we won't know they exist when they come from the registry).  When we build, we'll need to take into account the difference between the resolver graph and the unit graph, if we don't already.

In effect, we are making the parsing of the root crate in the `.crate` look as if nested packages don't exist.

Why
- Minimizes load on the Index and in dependency resolution because there aren't entries for these anonymous packages where there is no effect / no decisions need to be made
- Isolates the special cases to just building up the unit graph, rather than also dealing with it during resolution, by the registries, etc
- Makes things easy for registries that use the Summary or registries that recreate it from `Cargo.toml` like crates.io (no complex package parsing logic needed)

Reply:
I've added a paragraph on this to the RFC text: "The package index does not explicitly represent nested packages; instead, nested packages' dependencies are flattened into the dependencies of the parent package."